### DISCUSSION TITLE
Updating external atoms within an effect

### DISCUSSION THREAD

**Discussion by tpict** [facebookexperimental/Recoil#1428] [Q&A]
Say I want to add an effect to my `effects_UNSTABLE` array that does some asynchronous work when `myAtom` enters a particular state. I want to handle failure modes by displaying a popup warning to the user. Popup warnings are governed globally by `warningsAtom`.

I can't think of any solution that doesn't implement effects at the component level:
- write the effect in a `useEffect` where I can bring `warningsAtom` into scope
- hold warnings in `myAtom`, and write a `useEffect` to copy them to `warningsAtom`

but either way, it means a critical part of state management now lives outside Recoil.

Transactions looked like a promising solution, but it's a dead end without async support.

`useRecoilCallback` checks the boxes for "does async" and "can update multiple atoms", but you can't check to see if state updated while the async work was being done.

I'm guessing that `effects_UNSTABLE` doesn't expose a generic `set` method to encourage "atomic" atoms, but without an alternative it feels like a hole in the Recoil API.

What's the best way of implementing this?

**Comment by dnsco** [comment]
I hit this issue as well, and wound up moving all my logic to custom hooks :/

On the plus side I've found that testing everything at the hook level allows for moving the implementation between atom/react effects,  so I'm hopeful, but ultimately the hook versions wound up being a lot cleaner for me.

**Reply by tpict** [reply]
I thought about custom hooks. Say the side-effects are implemented in some custom hook that provides a callback:
```tsx
const useDoSomething = () => {
  const [state, setState] = useRecoilState(myAtom);
  const [warning, setWarning] = useRecoilState(warningsAtom);
  
  useEffect(() => {
    setWarning(`WARNING: State is now ${state}`);
  }, [state]);

  return () => setState(state => state + 1);
};

const MyComponent = function() {
  const doSomething = useDoSomething();

  return <button onClick={doSomething}>Click me</button>
}
```

Now I decide I want to use `useDoSomething` in a different component. This results in the effect being triggered twice for a single state shared state update.

This is a contrived example. The effect could be implemented inside the callback logic to avoid the issue, but that would be more-or-less equivalent to doing everything in `useRecoilCallback`.

Do you have an example of how you've composed your custom hooks?

**Comment by dnsco** [comment]
[Here's a custom hook](https://github.com/dnsco/better-pwa/blob/a4b5f07f77e5170ae66e9d411d37373358193276/src/state/useMyActivities.ts#L45). I'm using useRecoilCallback. [I worte a library](https://github.com/dnsco/recoil-test-render-hooks/) to help with testing– I've to  a bunch of tests around that hook I wrote in that repo. If you wanted to test that it only happens once you could inject a jest.mock, render it multiple times, and assert the mock only gets called once, to make sure you got the dependency stuff correct. TBH I'm not completely clear on the memoization/effect dependency stuff.

The use transaction stuff is a little sad because it doesn't work with selectors apparently. If you figure out a better way to do all this please let me know!

**Reply by tpict** [reply]
Thanks!

**Comment by tpict** [comment]
One small API tweak that would make this more manageable: allow `useRecoilCallback` to call its result, but receive the latest snapshot when it does so.

```tsx
const animalAtom = atom({ key: "animal", default: "cat" });
const infoAtom = atom({ key: "info", default: undefined });

type Action = 
  | { type: "CHANGE_ANIMAL"; animal: string }
  | { type: "UPDATE_ANIMAL_INFO"; animal: string; info: string; };

useRecoilCallback<[Action], Promise<void>>(({ snapshot, set, callSelf }) => async action => {
  switch (action.type) {
    case "CHANGE_ANIMAL":
      set(animalAtom, action.animal);
      asynchronouslyGatherInfo(action.animal).then(
        info => callSelf({ type: "UPDATE_ANIMAL_INFO", animal: action.animal, info })
      );
      return;
    case "UPDATE_ANIMAL_INFO":
      const animal = await snapshot.getPromise(animalAtom);
      if (animal !== action.animal) {
        // animal changed while our async job was in progress
        return;
      }

      set(infoAtom, action.info);
      return;
  }
});

**Comment by tpict** [comment]
I've only done preliminary testing of it, but I've come up with this:

```tsx
import { CallbackInterface, useRecoilCallback } from "recoil";

export interface AsyncDispatch<A> {
  (action: A): Promise<void>;
}

export interface DispatchWithThunk<A> {
  (action: A): Promise<void | ((dispatch: AsyncDispatch<A>) => void)>;
}

export function useRecoilDispatch<A>(
  fn: (callbackInterface: CallbackInterface) => DispatchWithThunk<A>,
): AsyncDispatch<A> {
  const dispatch: AsyncDispatch<A> = useRecoilCallback(callbackInterface => {
    const handler = fn(callbackInterface);
    return async action => {
      const thunk = await handler(action);
      if (typeof thunk === "function") {
        thunk(dispatch);
      }
    };
  }, []);

  return dispatch;
}
```

The example above then looks like

```tsx
const animalAtom = atom({ key: "animal", default: "cat" });
const infoAtom = atom({ key: "info", default: undefined });

type Action = 
  | { type: "CHANGE_ANIMAL"; animal: string }
  | { type: "UPDATE_ANIMAL_INFO"; animal: string; info: string; };

useRecoilDispatch<Action>(({ snapshot, set }) => async action => {
  switch (action.type) {
    case "CHANGE_ANIMAL":
      set(animalAtom, action.animal);
      return async dispatch => {
        const info = await asynchronouslyGatherInfo(action.animal);
        dispatch({ type: "UPDATE_ANIMAL_INFO", animal: action.animal, info });
      };
    case "UPDATE_ANIMAL_INFO":
      const animal = await snapshot.getPromise(animalAtom);
      if (animal !== action.animal) {
        // animal changed while our async job was in progress
        return;
      }

      set(infoAtom, action.info);
      return;
  }
});
```

(bears mentioning this is a contrived example and a better solution would be to make `infoAtom` an atom family parameterized by animal)

The thunk/dispatch terminology isn't entirely accurate but the API will look familiar to Redux users. I didn't set out to replicate it 1:1 but it's preferable to me over [the work-around for retrieving "current" state floated here](https://github.com/facebookexperimental/Recoil/issues/829#issuecomment-759067655), because the handler for each action should only receive the last valid snapshot... I think.

With this hook, I can guard state updates _and_ queue updates that require access to future state, which sufficiently eases the problem described in the OP.

Still, this approach requires discipline: each action handler you implement that updates `animalAtom` has to dispatch `UPDATE_ANIMAL_INFO` with fresh data. The effect pattern would be more maintainable, but neither implementation I know of fits the bill:

- Recoil effects can only modify their own atom.
- React's `useEffect` is tied to a component's lifecycle, so for this problem you'd have to write the effect way up at the top of the tree with `RecoilRoot`. If you're keeping all your state updates in one `useRecoilCallback`, it would be preferable to colocate the effects with that logic, but then your effects would be duplicated in each place you use the callback hook. Maybe you could solve this with another context provider beneath `RecoilRoot`?
